File: code/functions/pyprism/pyprism/data/_load_wes.py
# -*- coding: utf-8 -*-
"""
@created: Jan 24 2021
@modified: Apr 26 2021
@author: Yoann Pradat

    CentraleSupelec
    MICS laboratory
    9 rue Juliot Curie, Gif-Sur-Yvette, 91190 France

    Institut Gustave Roussy
    Prism Center
    114 rue Edouard Vaillant, Villejuif, 94800 France

Functions
    def _build_load_stat_for_wes_mut(df_maf, patient_field, tumor_field, normal_field):
    def load_wes_mut(study: str, identifiers: Iterable[str]=None, identifiers_name: str=None, mode:
        str="somatic_filtered"):
"""

import os
import pandas as pd
from typing import Iterable

from pyprism.util import setwd_to_data
from ._util_load import subset_data, load_from_data
from ._filepaths import _get_filepath_wes_mut

DataFrame = pd.core.frame.DataFrame

def _build_load_stat_for_wes_mut(df_maf, patient_field, tumor_field=None, normal_field=None):
    if tumor_field is None and normal_field is not None:
        df_stat = df_maf.loc[:, [patient_field, normal_field]].drop_duplicates()
        df_stat = df_stat.reset_index(drop=True)
        df_stat_gby = df_stat.groupby(patient_field)
        df_stat = pd.DataFrame({"N_Loaded": df_stat_gby[normal_field].apply(lambda x: x.nunique()),
                                normal_field: df_stat_gby[normal_field].apply(lambda x: "|".join(x))})
        df_stat = df_stat.reset_index(drop=False)

        message = "normal sample"
    else:
        df_stat = df_maf.loc[:, [patient_field, tumor_field, normal_field]].drop_duplicates()
        df_stat = df_stat.reset_index(drop=True)
        tumor_vs_normal = "%s_vs_%s" % (tumor_field, normal_field)
        df_stat.loc[:, tumor_vs_normal] = df_stat[[tumor_field, normal_field]].agg(lambda x: "_vs_".join(x), axis=1)
        df_stat = df_stat[[patient_field, tumor_vs_normal]].reset_index(drop=True)
        df_stat_gby = df_stat.groupby(patient_field)
        df_stat = pd.DataFrame({"N_Loaded": df_stat_gby[tumor_vs_normal].apply(lambda x: x.nunique()),
                                tumor_vs_normal: df_stat_gby[tumor_vs_normal].apply(lambda x: "|".join(x))})
        df_stat = df_stat.reset_index(drop=False)

        message = "couple(s) tumor vs normal"

    # print
    s_n_loaded = df_stat["N_Loaded"].value_counts()

    for n in sorted(s_n_loaded.index):
        if n <= 2:
            print("%d %s loaded for %d patient(s)" % (n, message, s_n_loaded[n]))
        else:
            print("%d %s loaded for %d patient(s)" % (n, message, s_n_loaded[n]))
            print("\t" + "\n\t".join(list(df_stat.loc[df_stat.N_Loaded == n, patient_field].values)))

    return df_stat

# User function ========================================================================================================

def load_wes_mut(study: str, identifiers: Iterable[str]=None, identifiers_name: str=None, mode: str="somatic_maf",
             return_load_stat: bool=False):
    """
    Load mutation tables for the specified study.

    Parameters
    ----------
    study: str
        Choose 'met500', 'tcga' or 'prism'.
    identifiers: Iterable[str], optional.
        If not None, return annotations only for the identifiers values.
    identifiers_name: Iterable[str], optional.
        Name of the column of identifiers.
    mode: str, default="somatic_maf"
        1. If "somatic_maf", load data/[study]/wes/somatic_maf/somatic_calls.maf.gz
        2. If "somatic_filters", load data/[study]/wes/somatic_maf/somatic_calls_filters.tsv.gz
        3. If "somatic_oncokb", load data/[study]/wes/somatic_maf/somatic_calls_oncokb.maf.gz
        4. If "somatic_civic", load data/[study]/wes/somatic_maf/somatic_calls_civic.maf.gz

    return_load_stat: bool, default=True
        If True, returns a tuple of dataframe with first the maf table and second a table providing details about what
        was loaded. If False, returns only the MAF table.


    Returns
    -------
    df_maf, df_stat: tuple of DataFrame
    """

    # load maf file
    print("-reading aggregated MAF file ... ", end="")
    df_maf = load_from_data(_get_filepath_wes_mut(study, mode), header_prefix="##",
                            sep="\t", dtype={"Chromosome": str}, low_memory=False)
    print("ok!")

    # # use metadata to add sample and subject identifiers
    # df_meta = load_summary_wes_mut(study, mode, use_cache=True)
    # del df_meta["N_Mutations"]
    # cols_com = list(set(df_meta.columns).intersection(set(df_maf.columns)))
    # df_maf = df_maf.merge(df_meta, how="left", on=cols_com)

    # subset if any
    df_maf = subset_data(df_maf, values=identifiers, col_name=identifiers_name)

    if return_load_stat:
        if "oncotator" in mode or "mc3" in mode:
            # build a table with one line per subject and the number of pairs tumor-normal loaded for that patient
            df_stat = _build_load_stat_for_wes_mut(df_maf, "Subject_Id",
                                               tumor_field="Tumor_Sample_Id", normal_field="Normal_Sample_Id")
        elif "annovar" in mode:
            df_stat = _build_load_stat_for_wes_mut(df_maf, "Subject_Id", normal_field="Sample_Id")

        return df_maf, df_stat
    else:
        return df_maf

File: code/functions/rprism/R/load_wes.R
# @created: 25 Jan 21
# @modified: 30 Aug 21
# @author: Yoann Pradat
# 
# Exported functions
#     load_stat_for_wes_mut
#     load_wes_mut

#' Build a dataframe with per patient statistics of MAF files loading.
#'
#' Used to see which data was loaded for which patient in MAF files.
#'
#' @return a list of \code{data.frame}
#'
#' @param df_maf The dataframe in MAF format
#' @param patient_field The name of the patient column
#' @param tumor_field The name of the tumor column
#' @param normal_field The name of the normal column
#'
#' @importFrom dplyr group_by summarize n
#' @importFrom rlang := .data
#' @importFrom magrittr %>%
#'
#' @author Yoann Pradat
#'
#' @export
load_stat_for_wes_mut <- function(df_maf, patient_field, tumor_field=NULL, normal_field=NULL){
  if (is.null(tumor_field) & !is.null(normal_field)){
    df_stat <- unique(df_maf[,c(patient_field, normal_field)])
    rownames(df_stat) <- seq(1, nrow(df_stat))

    df_stat <- df_stat %>%
      dplyr::group_by(.data[[patient_field]]) %>%
      dplyr::summarize(N_Loaded=dplyr::n(), {{ normal_field }}:=paste(.data[[normal_field]], collapse="|"))

    message <- "normal sample"
  } else {
    df_stat <- unique(df_maf[,c(patient_field, tumor_field, normal_field)])
    rownames(df_stat) <- seq(1, nrow(df_stat))

    tumor_vs_normal <- paste(tumor_field, normal_field, sep="_vs_")
    df_stat[, tumor_vs_normal] <- paste(df_stat[[tumor_field]], df_stat[[normal_field]], sep="_vs_")
    df_stat <- df_stat[, c(patient_field, tumor_vs_normal)]

    df_stat <- df_stat %>%
      dplyr::group_by(.data[[patient_field]]) %>%
      dplyr::summarize(N_Loaded=dplyr::n(), {{ tumor_vs_normal }}:=paste(.data[[tumor_vs_normal]], collapse="|"))

    message <- "couple(s) tumor vs normal"
  }

  # print
  s_n_loaded <- table(df_stat$N_Loaded)

  for (n in sort(names(s_n_loaded))){
    if (as.integer(n) <= 2){
      cat(paste(n, message, "loaded for", s_n_loaded[[n]], "patient(s)\n"))
    } else {
      cat(paste(n,  message, "loaded for", s_n_loaded[[n]], "patient(s)\n"))
      cat(paste0("\t", paste(df_stat[df_stat$N_Loaded == as.integer(n), patient_field, drop=T], collapse="\n\t"), "\n"))
    }
  }

  df_stat
}


#' Load mutation tables.
#'
#' Load mutation tables for the specified study and optionally specified identifiers. This function is an R
#' reimplementation of the function \code{load_wes_mut} from the pyprism package.
#'
#' \itemize{
#'  \item If "somatic_maf", load data/[study]/wes/somatic_maf/somatic_calls.maf.gz
#'  \item If "somatic_filters", load data/[study]/wes/somatic_maf/somatic_calls_filters.tsv.gz
#'  \item If "somatic_oncokb", load data/[study]/wes/somatic_maf/somatic_calls_oncokb.maf.gz
#'  \item If "somatic_civic", load data/[study]/wes/somatic_maf/somatic_calls_civic.maf.gz
#' }
#'
#' @return a \code{data.frame}
#'
#' @param study Name of the cohort used in the naming of the files. Choose one of "met500", "prism" or "tcga".
#' @param identifiers (optional) If not NULL, return clinical tables only for the individuals specified
#' @param identifiers_name (optional) The name of the column the \code{identifiers} values correspond to.
#' @param mode (optional) Choose which MAF file to load.
#' @param ... Extra parameters passed to \code{\link{load_from_data}}
#'
#' @importFrom readr cols
#' @importFrom dplyr left_join
#'
#' @author Yoann Pradat
#'
#' @export
load_wes_mut <- function(study, identifiers=NULL, identifiers_name=NULL, mode="somatic_maf", ...){
  # load maf file
  cat("-reading aggregated MAF file ... ")
  df_maf <- load_from_data(get_filepath_wes_mut(study, mode), header_prefix="##", col_types=readr::cols(), ...)
  cat("ok!\n")

  # use summary data to add sample and subject identifiers
  # df_summary <- load_summary_wes_mut(study, mode)
  # df_summary[, "N_Mutations"] <- NULL
  # cols_com <- intersect(colnames(df_summary), colnames(df_maf))
  # df_maf <- left_join(df_maf, df_summary, by=cols_com)

  # subset if any
  df_maf <- subset_data(df_maf, values=identifiers, col_name=identifiers_name)

  df_maf
}
